tools/base/aio.py

```diff
@@ -83,3 +105,405 @@ async def run_system_command():
+        await asyncio.sleep(wait)
+        return i, wait
+
+    async def run(coros):
+        async for (i, wait) in aio.concurrent(coros, limit=3):
+            print(f"{i} waited {wait}")
+
+    def provider():
+        for i in range(0, 10):
+            yield task_to_run(i)
+
+    asyncio.run(run(provider()))
+    ```
+    """
+
+    def __init__(
+            self,
+            coros: Union[types.AsyncGeneratorType, AsyncIterable[Awaitable],
+                         AsyncIterator[Awaitable], types.GeneratorType, Iterator[Awaitable],
+                         Iterable[Awaitable]],
+            yield_exceptions: Optional[bool] = False,
+            limit: Optional[int] = None):
+        self._coros = coros
+        self._limit = limit
+        self._running: List[asyncio.Task] = []
+        self.yield_exceptions = yield_exceptions
+
+    def __aiter__(self) -> AsyncIterator:
+        """Start a coroutine task to process the submit queue, and return
+        an async generator to deliver results back as they arrive
+        """
+        self.submit_task = asyncio.create_task(self.submit())
+        return self.output()
+
+    @property
+    def active(self) -> bool:
+        """Checks whether the iterator is active, either because it
+        hasn't finished submitting or because there are still tasks running
+        """
+        return self.submitting or self.running
+
+    @property
+    def closed(self) -> bool:
+        """If an unhandled error occurs, the generator is closed and no further
+        processing should happen
+        """
+        return self.closing_lock.locked()
+
+    @cached_property
+    def closing_lock(self) -> asyncio.Lock:
+        """Flag to indicate whether the generator has been closed"""
+        return asyncio.Lock()
+
+    @async_property
+    async def coros(self) -> AsyncIterator[Union[ConcurrentIteratorError, Awaitable]]:
+        """An async iterator of the provided coroutines"""
+        coros = self.iter_coros()
+        try:
+            async for coro in coros:
+                yield coro
+        except GeneratorExit:
+            # If we exit before we finish generating we land here (ie error was raised)
+            # In this case we need to tell the (possibly) async generating provider to
+            # also close.
+            try:
+                await coros.aclose()  # type:ignore
+            finally:
+                # Suppress errors closing the provider generator
+                # This can raise a further `GeneratorExit` but it will stop providing.
+                return
+
+    @property
+    def default_limit(self) -> int:
+        """Default is to use cpu+4 to a max of 32 coroutines"""
+        # This reflects the default for asyncio's `ThreadPoolExecutor`, this is a fairly
+        # arbitrary number to use, but it seems like a reasonable default.
+        return min(32, (os.cpu_count() or 0) + 4)
+
+    @cached_property
+    def is_async(self) -> bool:
```

Comment:
Nit: I found this naming confusing, would prefer something like `coros_are_async`. It's not the concurrent loop that is async or not, it's the coros.. same with the others.

Reply:
>  It's not the concurrent loop that is async or not, it's the coros.. same with the others.

hmm, this is diametrically wrong - im wondering how i can make this clearer

its not the coros, its the `generator` that provides the coros that is a/sync

Reply:
ill improve the comment for a start...

Reply:
One thing that is important to rem with this implementation is that it *is* an `AsyncGenerator` and it consumes an (optionally a/sync) generator (or any iterable in fact)

Reply:
not sure about good names - something that conveys `iterates_generatively` and `iterates_async`

Reply:
i think perhaps `consumes_generator` and `consumes_async` - ill update...